File: 2014/Qualification Round/A_Magic_Trick/MagicTrick.c
#include <stdio.h>
#include <stdlib.h>

#define ERR_BAD_MAGICIAN (-1)
#define ERR_VOLUNTEER_CHEAT (-2)

int findCard(int selectedRow1st[4], int selectedRow2nd[4])
{
    // incoming row data is sorted
    int matchNumber = 0;
    int i = 0, j = 0;
    while(i < 4 && j < 4) {
        if (selectedRow1st[i] == selectedRow2nd[j]) {
            if (matchNumber) {
                return ERR_BAD_MAGICIAN;
            }
            matchNumber = selectedRow1st[i];
            i++;
            j++;
        } else if (selectedRow1st[i] <= selectedRow2nd[j]) {
            i++;
        } else {
            j++;
        }
    }

    if (matchNumber) {
        return matchNumber;
    }

    return ERR_VOLUNTEER_CHEAT;
}

int compareInt(const void *first, const void *second)
{
    return (*(int *)first - *(int *)second);
}

void selectCard(/* out */ int rowData[4])
{
    int selectedRow, selectedRangeMin, selectedRangeMax;

    scanf("%d", &selectedRow);
    selectedRangeMin = (selectedRow - 1) * 4;
    selectedRangeMax = selectedRangeMin + 4;

    int cardNumber, *pRowData = rowData;
    for (int i = 0; i < 16; i++) {
        scanf("%d", &cardNumber);
        if (i >= selectedRangeMin && i < selectedRangeMax) {
            *pRowData = cardNumber;
            pRowData++;
        }
    }

    // sort rowData
    qsort(rowData, 4, sizeof(int), compareInt);
}

int main(int argc, char *argv[])
{
    int T;
    scanf("%d", &T);

    int selectedRow1st[4], selectedRow2nd[4], result;
    for (int i = 0; i < T; i++) {
        selectCard(selectedRow1st);
        selectCard(selectedRow2nd);
        result = findCard(selectedRow1st, selectedRow2nd);

        printf("Case #%d: ", i + 1);
        switch (result) {
        case ERR_BAD_MAGICIAN:
            printf("Bad magician!\n");
            break;
        case ERR_VOLUNTEER_CHEAT:
            printf("Volunteer cheated!\n");
            break;
        default:
            printf("%d\n", result);
            break;
        }
    }

    return 0;
}

File: 2014/Qualification Round/A_Magic_Trick/MagicTrick.py
import sys

def debug(x):
#    print(x)
    return

N = int(input())
debug('N: ' + str(N))

def getCards():
    R = int(input()) - 1
    debug('R: ' + str(R))
    result = []
    for i in range(4):
        if R != i:
            input()
        else:
            result = sorted([int(p) for p in input().split(' ')])
    debug("### getCards: " + ' '.join([str(r) for r in result]))
    return result

for n in range(N):
    card1 = getCards()
    card2 = getCards()
    answer = [c1 for c1 in card1 for c2 in card2 if c1 == c2]
    debug("### answer: " + ' '.join([str(r) for r in answer]))
    if len(answer) == 1:
        print('Case #%d: %d' % (n + 1, answer[0]))
    elif len(answer) == 0:
        print('Case #%d: Volunteer cheated!' % (n + 1))
    else:
        print('Case #%d: Bad magician!' % (n + 1))